### DISCUSSION TITLE
Some questions regarding security

### DISCUSSION THREAD

**Discussion by andreabadesso** [WalletConnect/walletconnect-monorepo#2230] [Developer Support]
From my investigation, `web3wallet` always uses the same `symKey` for sessions, it stores it in the storage (AsyncStorage on react-native or localStorage on web) in plain-text using the `wc@2:core:0.3//keychain` key

1. Is it possible to make this `symKey` ephemeral, meaning that it gets rotated from time to time (or after a number of messages exchanged)?
1. Is it possible to encrypt the `symKey` instead of storing it in plain-text?
1. I've noticed it [always uses](https://github.com/WalletConnect/walletconnect-monorepo/blob/557b9e295039285cdbab2e3183211095aee9e94a/packages/utils/src/crypto.ts#L63) `ChaCha20-Poly1305` for encrypting messages with the `symKey`, is it possible to use a different algorithm? Maybe a list of different algos the `dApp`/`App` can choose from?

I couldn't find any information on this in the docs

Thanks!!

**Comment by mcsherrylabs** [comment]
Hi Andre, apologies for the false alarm, I have no answers for you only more questions!    

I've been looking for documentation of the specifics of pairing (v2), generating the shared key and what auditing of that scheme has been done. It has to be out here somewhere, but I've had no luck. 

Your question makes me think you could point me in the right direction?

Actually I found this https://docs.walletconnect.com/2.0/specs/clients/core/pairing/ 

Do you know of any other doco or security reviews?

Any steer appreciated.

**Reply by andreabadesso** [reply]
Hi @mcsherrylabs! Sorry for the late response

I did some investigation [here](https://github.com/HathorNetwork/rfcs/blob/22655b8ab6795e99305c604343633776ac433c25/projects/web-wallet/wallet-connect/0001-design.md) by reading the wallet connect code, there is an explanation from what I could understand from their docs and code

Hope it helps

**Comment by finessevanes** [comment]
Passed along your questions to our team!

**Comment by llbartekll** [comment]
Hey @andreabadesso why you’d like to use different algorithm?

**Reply by andreabadesso** [reply]
I don't necessarily want to use a different algorithm, it would be nice to have the ability to do so though

The other points are the most relevant, would you mind taking a look at them as well?

**Comment by finessevanes** [comment]
>Is it possible to make this symKey ephemeral, meaning that it gets rotated from time to time (or after a number of messages exchanged)?

As it currently stands, key rotation doesn't occur for either identity keys or encryption keys. Adding rotation for the latter could lead to unnecessary complexity.

>Is it possible to encrypt the symKey instead of storing it in plain-text?

The symKey is encrypted across all platforms, which includes Android, Web, and iOS.

>I've noticed it [always uses](https://github.com/WalletConnect/walletconnect-monorepo/blob/557b9e295039285cdbab2e3183211095aee9e94a/packages/utils/src/crypto.ts#L63) ChaCha20-Poly1305 for encrypting messages with the symKey, is it possible to use a different algorithm? Maybe a list of different algos the dApp/App can choose from?

WalletConnect SDKs exclusively use ChaCha20-Poly1305 for encryption on all platforms. We currently do not support "algo" negotiation within the protocol.

**Comment by andreabadesso** [comment]
> The symKey is encrypted across all platforms, which includes Android, Web, and iOS.

Is it really? So maybe there is a bug because this:
`const symKeys = await AsyncStorage.getItem('wc@2:core:0.3//keychain');`

returns a key->value object with the sessionId and its `symKey`, in **plain-text**:

```
{
	"client_ed25519_seed": "afea6192caaeb0a4d3946706a0aed5259aa211bcd6e768a0b1068850d0b41f29",
	"2b07b3a4e70fd4450aa1ad316647f9a9ffa271d6e6e8b8c0d6082f09e3bb455f": "2fec9902de975cc6eb02617aea4bd83a5042efafa7de234f261feddc47b659e9",
}
```
